### DISCUSSION TITLE
Can the simscape be connected with python directly?

### DISCUSSION THREAD

**Discussion by lk1983823** [johbrust/simulink_gym#10] [Q&A] [answered]
I want to use python code to set values to a simscape block. For example, the Rational Damper block has Parameters `Damping coeffiecent `and Variables `Torque` and` Rational velocity`, all of which also have units to set. Can we assgin these values or units from python? The simscape is a physical model building block.
Thanks.

**Comment by johbrust** [comment, marked as answer]
I looked into this and found that this was not possible. Most (all?) Simscape block values can only be accessed via the GUI. But you can use a workspace variable for the block parameter values and change those programmatically. Therefore, I updated the definition of the observation space to also allow for the use of workspace variables for the initial values. Check out the [Simscape implementation of the cart pole environment](https://github.com/johbrust/simulink_gym/tree/main/examples/envs/cartpole_simscape) for that!

**Reply by lk1983823** [reply]
> I looked into this and found that this was not possible. Most (all?) Simscape block values can only be accessed via the GUI. But you can use a workspace variable for the block parameter values and change those programmatically. Therefore, I updated the definition of the observation space to also allow for the use of workspace variables for the initial values. Check out the [Simscape implementation of the cart pole environment](https://github.com/johbrust/simulink_gym/tree/main/examples/envs/cartpole_simscape) for that!

Thanks for the wonderful update. I run it successfully. And I noticed you have modified some of the orginal code. For example, BlockParam is removed and Observation's __init__ has changed its arugments. So I try to update the simple_reacher_simulink.py code as well. Can you try to run it and give me a verification? Thanks.

```
from simulink_gym import SimulinkEnv, Observation, Observations
from gym.spaces import Box
from pathlib import Path
import numpy as np
import math


class SimpleReacherSimulink(SimulinkEnv):
    def __init__(
        self,
        stop_time: float = 10.0,
        step_size: float = 0.02, 
        length_1=1.0, 
        length_2=0.5,
        model_debug: bool = False
        ):

        super().__init__(
            model_path=Path(__file__)
            .parent.absolute()
            .joinpath("simple_reacher_simulink.slx"),
            model_debug=model_debug,
        )

        self.set_block_parameter(
            f"{self.env_name}/length_1/Value", length_1)
        self.set_block_parameter(
            f"{self.env_name}/length_2/Value", length_2)

        # Define action space:
        self.action_space = Box(low=0.0, high=math.pi, shape=(2,), dtype=np.float32)

        # Define observation space:
        self.observations = Observations(
            [
                Observation(
                    "x",
                    -np.inf,
                    np.inf,
                    f"{self.env_name}/IC_1/Value",
                    self.set_block_parameter,
                ),
                Observation(
                    "y",
                    -np.inf,
                    np.inf,
                    f"{self.env_name}/IC_2/Value",
                    self.set_block_parameter,
                ),
            ]
        )

        # Get initial state from defined observations:
        self.state = self.observations.initial_state
        # Set simulation parameters:
        self.set_model_parameter("StopTime", stop_time)
        self.set_workspace_variable("step_size", step_size)

    def reset(self):
        # Call common reset:
        super()._reset()

        # Return reshaped state. Needed for use as tf.model input:
        return self.state

    def step(self, action):
        state, simulation_time, terminated, truncated = self.sim_step(action)

        # Check all termination conditions:
        done = bool(terminated or truncated)

        # Receive reward for every step inside state and time limits:
        reward = 1

        info = {"simulation time [s]": simulation_time}

        return state, reward, done, info

```

In addition, I can't run the cartploe example (both for simulink and simscape) under a debug mode. In the ipynb file, when I run `env = CartPoleSimulink(stop_time=30, model_debug=True)`  and start the .slx file manually and click the run botton it shows:
```
Error evaluating registered method 'Start' of MATLAB S-Function 'stcpiprb' in 'cartpole_simulink/TCP//IP In/TCP//IP Receive'.
Caused by:
Unsuccessful open: Connection refused (Connection refused) 
```

**Reply by johbrust** [reply]
This code looks correct! I also updated the simple reacher branch just now.

**Reply by lk1983823** [reply]
> This code looks correct! I also updated the simple reacher branch just now.



Thanks. In addition, I can't run the cartploe example (both for simulink and simscape) under a debug mode. In the ipynb file, when I run env = CartPoleSimulink(stop_time=30, model_debug=True) and start the .slx file manually and click the run botton it shows:

```
Error evaluating registered method 'Start' of MATLAB S-Function 'stcpiprb' in 'cartpole_simulink/TCP//IP In/TCP//IP Receive'.
Caused by:
Unsuccessful open: Connection refused (Connection refused) 
```
Have you tried the dubg mode?

**Reply by johbrust** [reply]
The debug mode was, indeed, not working. I fixed that with the newest version, so it should work now! Also regarding the debug mode: I also updated the Readme with the right point in time for starting the model. Previously, the Readme stated that the simulation has to be started once the environment was initialized. It works more reliably if you start the simulation while the initial reset is running. While resetting the connection is established, so the reset waits for the connection.

Sorry for all the bugs (this is still work in progress)! And THANK YOU very much for your very valuable input! You using and testing this is very much appreciated! Thanks!

**Reply by lk1983823** [reply]
> The debug mode was, indeed, not working. I fixed that with the newest version, so it should work now! Also regarding the debug mode: I also updated the Readme with the right point in time for starting the model. Previously, the Readme stated that the simulation has to be started once the environment was initialized. It works more reliably if you start the simulation while the initial reset is running. While resetting the connection is established, so the reset waits for the connection.
> 
> Sorry for all the bugs (this is still work in progress)! And THANK YOU very much for your very valuable input! You using and testing this is very much appreciated! Thanks!

Thanks. With pleasure!
